### PR TITLE
Reuse the epoll object to fix the dispatcher performance downgrade

### PR DESCRIPTION
**Pull request by interma**
In commit https://github.com/greenplum-db/gpdb/commit/e30909fccfef42a99070f080b7c06439d6d79b29, I used the "WaitEventSet" to refactor the dispatch logic.
But in the following pgbench test, we found the performance downgrade 50% due to this commit. Did a perf on it, the bottleneck is `native_queued_spin_lock_slowpath`, means there is a lot of lock contention in it. 

After investigation, we found the culprit is a large number of concurrent creation and destruction of epoll objects. And the fix idea is simple: reuse the epoll object.

Opened this PR to fix it, and:
* no need to backport to 6x, because it doesn't use WaitEventSet.
* no special test for it, the existing tests cover the whole changes.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [x] Review a PR in return to support the community

### CODE REVIEW

**Comment by interma** [comment]
I will attach some test results later.

File: src/backend/cdb/dispatcher/cdbgang.c [hunk 1]
```diff
@@ -94,6 +97,23 @@ cdbgang_createGang(List *segments, SegmentType segmentType)
 	return pCreateGangFunc(segments, segmentType);
 }
 
+/*
+ * init DispWaitSet: create it in the first time and reset it later
+ */
+void
+initDispatchWaitEventSet(int nevents)
```
Comment (adam8157) [inline src/backend/cdb/dispatcher/cdbgang.c:104]: How about moving this to `latch.c`? And renaming the function and taking the WaitSet and context as parameters.

Besides DRY, taking the WaitSet as a parameter would help others understand the codes.

File: src/backend/cdb/motion/ic_udpifc.c [hunk 2]
```diff
@@ -3705,35 +3707,35 @@ receiveChunksUDPIFC(ChunkTransportState *pTransportStates, ChunkTransportStateEn
 			nevent += nFds;
 	}
 
-	/* init WaitEventSet */
-	waitset = CreateWaitEventSet(CurrentMemoryContext, nevent);
+	/* reset WaitEventSet */
+	if (ICWaitSet == NULL)
```
Comment (adam8157) [inline src/backend/cdb/motion/ic_udpifc.c:3711]: DRY.

File: src/backend/storage/ipc/latch.c [hunk 3]
```diff
@@ -618,6 +618,80 @@ CreateWaitEventSet(MemoryContext context, int nevents)
...
+
+	/* the same alloc logic as CreateWaitEventSet() */
+	char	   *data;
+	Size		sz = 0;
+	sz += MAXALIGN(sizeof(WaitEventSet));
+	sz += MAXALIGN(sizeof(WaitEvent) * nevents);
+	sz += MAXALIGN(sizeof(struct epoll_event) * nevents);
+	data = (char *) MemoryContextAllocZero(context, sz);
+
+	/* reuse the epoll object and free the old memory */
```
Comment (adam8157) [inline src/backend/storage/ipc/latch.c:669]: Nitpick: the last `data +=` is not needed, set `set` as `data` at the beginning and set `pset` as `set` at the end for easy understanding.

```c
    /* reuse the epoll object and free the old memory */
    int save_epoll_fd = set->epoll_fd;
    pfree(set);

    set = (WaitEventSet *) data;
    data += MAXALIGN(sizeof(WaitEventSet));
    set->events = (WaitEvent *) data;
    data += MAXALIGN(sizeof(WaitEvent) * nevents);
    set->epoll_ret_events = (struct epoll_event *) data;

    set->latch = NULL;
    set->nevents_space = nevents;
    set->exit_on_postmaster_death = false;

    /* reuse the epoll object (is empty now) */
    set->epoll_fd = save_epoll_fd;
    *pset = set;

    return;
```

**Review by adam8157 — COMMENTED** [review]
Have you compared the performance with the version calling `poll()` directly?

**Comment by interma** [comment]
Thanks Adam.
> Have you compared the performance with the version calling `poll()` directly?

Yes, it's the "revert" version (means revert waiteventset to poll()) in my previous results, paste it again:
```
// revert e30909f
...
tps = 151641.552024 (including connections establishing)
tps = 151757.591140 (excluding connections establishing)
```

**Comment by interma** [comment]
And I will follow up the review advices after @QingMa-gp finish the review.

File: src/backend/cdb/motion/ic_udpifc.c (hunk 2 above)
Comment (interma) [reply src/backend/cdb/motion/ic_udpifc.c:3711]: got it, thanks.
I will fix both of them.

File: src/backend/storage/ipc/latch.c (hunk 3 above)
Comment (interma) [reply src/backend/storage/ipc/latch.c:669]: thanks, fixed

**Comment by interma** [comment]
@adam8157 all are resolved, please help to review again.

**Review by adam8157 — APPROVED** [review]

**Pull request merged**